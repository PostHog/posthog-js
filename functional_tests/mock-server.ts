// An MSW server that handles requests to https://app.posthog.com/e/ and stores

import { rest } from 'msw'
import { setupServer } from 'msw/lib/node'
import assert from 'assert'

// the request bodies in a store that we can inspect within tests.
const capturedRequests: { '/e/': any[]; '/engage/': any[]; '/decide/': any[] } = {
    '/e/': [],
    '/engage/': [],
    '/decide/': [],
}

const server = setupServer(
    rest.post('http://localhost/e/', (req: any, res: any, ctx: any) => {
        const body = req.body
        // type guard that body is a string
        if (typeof body !== 'string') {
            assert(false, 'body is not a string')
            return
        }
        // we assume the body is JSON and parse it we store the parsed body in
        // the store The request body is url encoded, so we need to decode it
        // first. We then need to get the data param from this and base64 decode
        // it.
        const data = JSON.parse(Buffer.from(decodeURIComponent(body.split('=')[1]), 'base64').toString())
        capturedRequests['/e/'] = [...(capturedRequests['/e/'] || []), data]
        return res(ctx.status(200))
    }),
    rest.post('http://localhost/engage/', (req: any, res: any, ctx: any) => {
        // NOTE: this is a slight duplication of the /e/ handler, but it's
        // possible that the details are slightly different so I'm leaving it
        // as is for now.
        const body = req.body
        // type guard that body is a string
        if (typeof body !== 'string') {
            assert(false, 'body is not a string')
            return
        }
        // we assume the body is JSON and parse it we store the parsed body in
        // the store The request body is url encoded, so we need to decode it
        // first. We then need to get the data param from this and base64 decode
        // it.
        const data = JSON.parse(Buffer.from(decodeURIComponent(body.split('=')[1]), 'base64').toString())
        capturedRequests['/engage/'] = [...(capturedRequests['/engage/'] || []), data]
        return res(ctx.status(200))
    }),
    rest.post('http://localhost/decide/', (req: any, res: any, ctx: any) => {
        // NOTE: this is a slight duplication of the /e/ handler, but it's
        // possible that the details are slightly different so I'm leaving it
        // as is for now.
        const body = req.body
        // type guard that body is a string
        if (typeof body !== 'string') {
            assert(false, 'body is not a string')
            return
        }
        // we assume the body is JSON and parse it we store the parsed body in
        // the store The request body is url encoded, so we need to decode it
        // first. We then need to get the data param from this and base64 decode
        // it.
        const data = JSON.parse(Buffer.from(decodeURIComponent(body.split('=')[1]), 'base64').toString())
        capturedRequests['/decide/'] = [...(capturedRequests['/decide/'] || []), data]
        return res(ctx.status(200), ctx.json({}))
    })
)

// Start the server before all the tests run, and close it afterwards.
beforeAll(() =>
    server.listen({
        onUnhandledRequest: 'error',
    })
)
afterAll(() => server.close())

export const getRequests = (token: string) => {
    // Filter the captured requests by the given token.
    return {
        '/e/': capturedRequests['/e/'].filter((request) => request.properties.token === token),
        '/engage/': capturedRequests['/engage/'].filter((request) => request.properties.token === token),
        '/decide/': capturedRequests['/decide/'].filter((request) => request.token === token),
    }
}

export const resetRequests = (token: string) => {
    Object.assign(capturedRequests, {
        '/e/': (capturedRequests['/e/'] = capturedRequests['/e/'].filter(
            (request) => request.properties.token !== token
        )),
        '/engage/': (capturedRequests['/engage/'] = capturedRequests['/engage/'].filter(
            (request) => request.properties.token !== token
        )),
        '/decide/': (capturedRequests['/decide/'] = capturedRequests['/decide/'].filter(
            (request) => request.token !== token
        )),
    })
}
