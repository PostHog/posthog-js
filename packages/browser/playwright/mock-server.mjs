/* eslint-env node */
import express from 'express'
import path from 'path'
import { createGunzip } from 'zlib'
import { pipeline } from 'stream'

const app = express()
const port = process.argv.includes('--port') ? parseInt(process.argv[process.argv.indexOf('--port') + 1]) : 2345
const eventStore = []

app.use('/static', express.static(path.resolve(process.cwd(), 'dist')))
app.use('/playground', express.static(path.resolve(process.cwd(), 'playground')))

app.get('/api/projects/1/events', (req, res) => {
    const properties = req.query.properties
    let matchingEvents = eventStore
    if (properties) {
        try {
            const parsedProperties = JSON.parse(properties)
            const testSessionId = parsedProperties.find((prop) => prop.key === 'testSessionId')
            if (testSessionId) {
                matchingEvents = eventStore.filter((event) => event.properties.testSessionId === testSessionId.value[0])
            }
        } catch (error) {
            console.error('Error parsing properties:', error)
        }
    }

    res.set({
        'Access-Control-Allow-Headers': '*',
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Credentials': 'true',
    }).json({ results: matchingEvents })
})

// Middleware to decompress gzip-encoded request bodies
app.use((req, res, next) => {
    if (req.query.compression === 'gzip-js') {
        const gunzip = createGunzip()
        let body = Buffer.concat([])

        pipeline(req, gunzip, (err) => {
            if (err) {
                console.error('Error decompressing gzip body:', err)
                return next(err)
            }
        })

        gunzip.on('data', (chunk) => {
            body = Buffer.concat([body, chunk])
        })

        gunzip.on('end', () => {
            try {
                req.body = JSON.parse(body.toString())
                next()
            } catch (parseError) {
                console.error('Error parsing decompressed body:', parseError)
                next(parseError)
            }
        })
    } else {
        next()
    }
})
app.use(express.json())
app.use(express.urlencoded({ extended: true }))

app.use(/^\/e/, (req, res) => {
    if (req.body) {
        let events
        if (Array.isArray(req.body)) {
            events = req.body.map((event) => event.event)
        } else {
            events = [req.body.event]
        }
        events.forEach((evt) => console.log(`[Event] ${evt}`))
        eventStore.push(req.body)
    } else {
        console.error('Received event request without body', req.url)
    }
    res.status(200)
        .set({
            'Content-Type': 'application/json',
            'Access-Control-Allow-Headers': '*',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Credentials': 'true',
        })
        .send(JSON.stringify({ status: 1 }))
})

app.use(/^\/ses/, (req, res) => {
    res.status(200)
        .set({
            'Content-Type': 'application/json',
            'Access-Control-Allow-Headers': '*',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Credentials': 'true',
        })
        .send(JSON.stringify({ status: 1 }))
})

app.listen(port, () => {
    console.log(`Mock server is running on http://localhost:${port}`)
}).on('error', (error) => {
    console.error(`Error starting mock server: ${error.message}`)
})
