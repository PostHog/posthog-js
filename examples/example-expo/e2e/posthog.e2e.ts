import { by, element, waitFor } from 'detox'
import jestExpect from 'expect'
import { createMockServer, MockRequest } from './mock-server'
const { reloadApp } = require('detox-expo-helpers')
const { objectContaining, arrayContaining } = jestExpect
const wait = (t: number) => new Promise((r) => setTimeout(r, t))

// Weird typescript issue
const toMatchSnapshot = (e: any) => (jestExpect(e) as any).toMatchSnapshot()

const commonProperties = {
  $app_build: '1',
  $app_name: 'Expo Go',
  $app_namespace: 'host.exp.Exponent',
  $app_version: '2.24.3',
  $device_manufacturer: 'Apple',
  $device_name: 'iPhone 12 Pro',
  $lib: 'posthog-react-native',
  $lib_version: jestExpect.any(String),
  $locale: jestExpect.any(String),
  $os_name: 'iOS',
  $os_version: '15.5',
  $screen_height: 844,
  $screen_width: 390,
  $timezone: jestExpect.any(String),
}
describe('PostHog React Native E2E', () => {
  let server: any
  let httpMock: jest.Mock<MockRequest, any>

  beforeAll(async () => {
    ;[server, httpMock] = createMockServer()
  })

  beforeEach(async () => {
    await reloadApp()
    httpMock.mockReset()

    await waitFor(element(by.id('title-TabOne')))
      .toBeVisible()
      .withTimeout(5000)
  })

  afterAll(async () => {
    await server.close()
  })

  it('should track $screen', async () => {
    await wait(1500)

    const calls = httpMock.mock.calls
    const eventCall = calls.find((x) => x[0].path === '/e/')

    jestExpect(eventCall[0]).toMatchObject({
      body: {
        api_key: 'phc_FzKQvNvps9ZUTxF5KJR9jIKdGb4bq4HNBa9SRyAHi0C',
        batch: arrayContaining([
          objectContaining({
            distinct_id: jestExpect.any(String),
            event: 'Application Opened',
            library: 'posthog-react-native',
            properties: {
              ...commonProperties,
            },
            timestamp: jestExpect.any(String),
            type: 'capture',
          }),
          objectContaining({
            distinct_id: jestExpect.any(String),
            event: '$screen',
            library: 'posthog-react-native',
            properties: {
              ...commonProperties,
              $screen_name: 'TabOne',
            },
            timestamp: jestExpect.any(String),
            type: 'capture',
          }),
        ]),
        sent_at: jestExpect.any(String),
      },
      method: 'POST',
      path: '/e/',
    })
  })

  it('should automatically track $screen on navigation', async () => {
    await wait(1500)
    httpMock.mockReset()

    await element(by.id('modal-button')).tap()
    await waitFor(element(by.id('title-Modal')))
      .toHaveLabel('Modal')
      .withTimeout(5000)

    await wait(1500)

    jestExpect(httpMock).toHaveBeenCalledWith(
      objectContaining({
        path: '/e/',
        body: objectContaining({
          batch: arrayContaining([
            objectContaining({
              event: '$screen',
              properties: objectContaining({
                $screen_name: 'Modal',
              }),
            }),
          ]),
        }),
      })
    )
  })

  it('should autocapture taps', async () => {
    await wait(1500)

    httpMock.mockReset()

    await element(by.id('example-ph-label')).tap()

    await wait(5000)

    const lastCall = httpMock.mock.lastCall
    jestExpect(lastCall[0]).toMatchObject({
      path: '/e/',
      body: objectContaining({
        batch: arrayContaining([
          objectContaining({
            event: '$autocapture',
            properties: objectContaining({
              $event_type: 'touch',
              $lib: 'posthog-react-native',
              $screen_height: 844,
              $screen_width: 390,
              $touch_x: jestExpect.any(Number),
              $touch_y: jestExpect.any(Number),
            }),
          }),
        ]),
      }),
    })
    toMatchSnapshot(lastCall[0].body.batch[0].properties.$elements)
  })

  it('should ignore autocapture for ph-no-capture', async () => {
    await wait(1500)
    httpMock.mockReset()

    await element(by.id('example-ph-no-capture')).tap()
    await wait(1000)

    jestExpect(httpMock).toHaveBeenCalledTimes(0)
  })
})
