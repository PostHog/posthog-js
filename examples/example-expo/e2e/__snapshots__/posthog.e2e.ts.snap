// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostHog React Native E2E should autocapture taps 1`] = `
Array [
  Object {
    "$el_text": "I have the property \\"ph-label\\" which means touching me will be autocaptured with a specific label",
    "attr__accessible": true,
    "attr__allowFontScaling": true,
    "attr__ellipsizeMode": "tail",
    "attr__isHighlighted": false,
    "attr__ph-label": "special-text",
    "attr__style": "",
    "attr__testID": "example-ph-label",
    "tag_name": "special-text",
  },
  Object {
    "$el_text": "I have the property \\"ph-label\\" which means touching me will be autocaptured with a specific label",
    "attr__ph-label": "special-text",
    "attr__testID": "example-ph-label",
    "tag_name": "special-text",
  },
  Object {
    "attr__style": "backgroundColor:rgba(0,0,0,.1);padding:15;borderRadius:10;marginBottom:10",
    "tag_name": "View",
  },
  Object {
    "tag_name": "View",
  },
  Object {
    "attr__style": "flex:1",
    "tag_name": "ScrollView",
  },
  Object {
    "attr__style": "flex:1",
    "tag_name": "ScrollView",
  },
  Object {
    "tag_name": "PostHogDemoScreen",
  },
  Object {
    "attr__name": "TabOne",
    "tag_name": "StaticContainer",
  },
  Object {
    "tag_name": "EnsureSingleNavigator",
  },
  Object {
    "tag_name": "SceneView",
  },
  Object {
    "attr__style": "flex:1",
    "tag_name": "View",
  },
  Object {
    "attr__accessibilityElementsHidden": false,
    "attr__importantForAccessibility": "auto",
    "attr__style": "flex:1;backgroundColor:rgb(242, 242, 242);flexDirection:column-reverse",
    "tag_name": "View",
  },
  Object {
    "attr__accessibilityElementsHidden": false,
    "attr__importantForAccessibility": "auto",
    "attr__style": "flex:1;flexDirection:column-reverse",
    "tag_name": "Background",
  },
  Object {
    "attr__focused": true,
    "attr__style": "",
    "tag_name": "Screen",
  },
  Object {
    "attr__activityState": 2,
    "attr__style": "position:absolute;left:0;right:0;top:0;bottom:0;zIndex:0",
    "tag_name": "AnimatedComponent",
  },
  Object {
    "attr__freeze": false,
    "tag_name": "MaybeFreeze",
  },
  Object {
    "attr__activityState": 2,
    "attr__enabled": true,
    "attr__style": "position:absolute;left:0;right:0;top:0;bottom:0;zIndex:0",
    "tag_name": "Screen",
  },
  Object {
    "attr__enabled": true,
    "attr__style": "position:absolute;left:0;right:0;top:0;bottom:0;zIndex:0",
    "attr__visible": true,
    "tag_name": "MaybeScreen",
  },
  Object {
    "attr__enabled": true,
    "attr__hasTwoStates": true,
    "attr__style": "flex:1;overflow:hidden",
    "tag_name": "ScreenContainer",
  },
  Object {
    "attr__enabled": true,
    "attr__hasTwoStates": true,
    "attr__style": "flex:1;overflow:hidden",
    "tag_name": "MaybeScreenContainer",
  },
]
`;
